### PR TITLE
Bug fixes in profiling allocator

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45993 Bug fixes in profiling allocator**

Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

### CODE REVIEW

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 1]
```diff
@@ -49,6 +31,36 @@ struct MemEvent {
...
+  for (const auto& event : alloc_events) {
+    auto alloc_id = event.allocation_id;
+    // Skip allocations not managed by AllocationPlan
+    if (allocation_offsets[alloc_id] == std::numeric_limits<uint64_t>::max()) {
+      continue;
+    }
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    if (event.type == EventType::Allocate) {
+      if (!allocations.emplace(start_offset, end_offset).second) {
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:47]: if you really want to validate - you should check that the inserted block doesn't overlap with existing blocks here

Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:47]: @dzhulgakov Thats what this emplace is supposed to do. Did I miss something?

Comment (dzhulgakov) [reply c10/mobile/CPUProfilingAllocator.cpp:47]: Emplace will only complain if the object is equal (which in this case is probably expressed via `!(a<b) && !(b<a)`). I originally assumed that you just sort the items in the set using their `start_offsets`, but it looks like you actually define the partial order above: `    return end_offset <= other.start_offset;	`.

I think it's technically not correct to create `std::set` with such comparison operator. STL documentation requires "strict weak ordering" (http://www.cplusplus.com/reference/set/set/), i.e. https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings . In particular string weak ordering requires "transitivity of incompatibility":

> For all x, y, z in S, if x is incomparable with y (neither x < y nor y < x hold), and y is incomparable with z, then x is incomparable with z (transitivity of incomparability).

However, you operator doesn't satisfy it, it's a non-strict weak ordering, e.g.:

```
x = (1, 10)
y = (5, 15)
z = (11, 20)
x < y || y < x == False # x and y are incompatible
y < z || z < y == False # y and z are incompatible
x < z # boom!
```

What it means that under certain overlapping circumstances you can get crash/assertion/runtime error being thrown from std::set operation.

I'd also give you another reason to not do it: it's very non-obvious to rely on such fancy comparison operator and it's very likely the reader of this code will overlook it (like I did).

To summarize: please change comparison operator to just look at starts only (so that it's a strict ordering) and then add necessary checks here that the inserted element doesn't overlap with the existing ones. It should be sufficient to just check the iterator just after the inserted one and verify that its start is after the inserted end.

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 1 above)
Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:47]: @dzhulgakov, thanks for pointing this out. I did miss "strict weak ordering" part and thanks for the explanation. That does make sense. Also for readability part, I do agree. It does get a bit cryptic. I should have put comments there at the least.

I will make changes.

Thanks once again. This was informative.

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 2]
```diff
@@ -43,12 +51,33 @@ bool validate_allocation_plan(
...
+      //    i.e. first_it.start_offset >= mem_block.end_offset
+      //    all blocks after this iterator dont overlap with mem_block
+      // 3. Iterate over all blocks from the beginning till iterator and check
+      //    for overlaps.
+      // Still O(n)
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
+        while(it != allocations.begin()) {
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:68]: you don't need to iterate actually. You just need to check two blocks - one just before (`--it`) and the one just after (`it`). Because the blocks in the set are note overlapping, it's sufficient

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 3]
```diff
@@ -49,6 +31,65 @@ struct MemEvent {
...
+          }
+          --it;
+        }
+        if (overlaps(*it, mem_block)) {
+          return false;
+        }
+      }
+      allocations.emplace(mem_block);
+    } else if (event.type == EventType::Free) {
+      auto it = allocations.find(mem_block);
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:80]: nit: you also need to validate the end of the range here. Because of your comparison operator `find` only finds the element based on the start

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 4]
```diff
@@ -49,6 +31,65 @@ struct MemEvent {
...
+      //    is not less than, i.e. >=, of this block.
+      //    i.e. first_it.start_offset >= mem_block.end_offset
+      //    all blocks after this iterator dont overlap with mem_block
+      // 3. Iterate over all blocks from the beginning till iterator and check
+      //    for overlaps.
+      // Still O(n)
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:67]: you still need to check the previous block even if it IS the end (as long as it's not the beginning)

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 2 above)
Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:68]: Yes thats right. I think I forgot the invariant that this set will always only contain non-overlapping intervals. With that invariant you need to check maximum of two other intervals.

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 3 above)
Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:80]: So based on the previous comment regarding the invariant that the set contains only non-overlapping intervals, check based on start offset should suffice. And given allocations should always precede free, if we have found overlapping interval we would have returned from this function beforehand.
Although I might have missed something again.

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 4 above)
Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:67]: Thanks for that catch.

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 2 above)
Comment (kimishpatel) [reply c10/mobile/CPUProfilingAllocator.cpp:68]: Also just a clarifying comment: candidate_block = (some_start, some_end) while the block I was using to perform query was (end_offset, max) so the search will point you the first block with start_offset >= some_end. However give the other comment regarding invariant, this does not matter anymore.

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 5]
```diff
@@ -49,6 +31,63 @@ struct MemEvent {
...
+    // Skip allocations not managed by AllocationPlan
+    if (allocation_offsets[alloc_id] == std::numeric_limits<uint64_t>::max()) {
+      continue;
+    }
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    MemBlock mem_block(start_offset, end_offset);
+    if (event.type == EventType::Allocate) {
+      auto it = allocations.lower_bound(mem_block);
+      if (!allocations.empty()) {
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:57]: you don't need this if statement

File: c10/mobile/CPUProfilingAllocator.cpp [hunk 6]
```diff
@@ -49,6 +31,63 @@ struct MemEvent {
...
+        if (it != allocations.end()) {
+          auto next_block = *it;
+          if (overlaps(next_block, mem_block)) {
+            return false;
+          }
+          if (it != allocations.begin()) {
```
Comment (dzhulgakov) [inline c10/mobile/CPUProfilingAllocator.cpp:63]: just pull it out of the surrounding `if` statement on `end` - your `else` part does exactly the same thing

File: c10/mobile/CPUProfilingAllocator.cpp (hunk 3 above)
Comment (dzhulgakov) [reply c10/mobile/CPUProfilingAllocator.cpp:78]: I wasn't referring to overlaps here, I was referring to sanity checking that the end of the interval you have in the set is the same as the end of the interval you're trying to remove.

**Review by dzhulgakov — APPROVED** [review]
Please address comments before landing

**Comment by kimishpatel** [comment]
@dzhulgakov Can you also approve the diff: D24172599
